firedrake/constant.py

```diff
@@ -14,19 +15,24 @@
 __all__ = ['Constant']
 
 
-def _globalify(value, comm):
+def _create_dat(op2type, value, comm):
+    # ~ if op2type is op2.Literal and not comm is None:
+        # Literals raise value error if created with a comm
+    if op2type is op2.Global and comm is None:
+        raise ValueError("Attempted to create pyop2 Global with no communicator")
+
     data = np.array(value, dtype=ScalarType)
     shape = data.shape
     rank = len(shape)
     if rank == 0:
-        dat = op2.Global(1, data, comm=comm)
+        dat = op2type(1, data, comm=comm)
     else:
-        dat = op2.Global(shape, data, comm=comm)
+        dat = op2type(shape, data, comm=comm)
     return dat, rank, shape
 
 
-class Constant(ufl.Coefficient, ConstantMixin):
-
+# Think "literal"
+class Constant(ufl.constantvalue.ConstantValue, ConstantMixin):
```

Comment:
Surely the right thing to inherit from is https://github.com/FEniCS/ufl/blob/d2dfbef9a60bede67ac8361712316bf427f9eecc/ufl/constant.py#L18

?

`ConstantValue` is a compile-time literal in UFL, not a runtime constant.

Reply:
The reason we have done this is that a UFL `Constant` is what Firedrake calls a `Function` in R. This means that it has an associated `ufl_domain` and therefore a communicator. The reason for the changes being made here is that many people use `Constant` just for parametrising kernel arguments and in some cases (like defcon) finding an appropriate mesh/communicator is tricky and we get deadlocks.

Our solution, not quite shown here, is for TSFC to define a `KernelArgument` mixin class that marks an element of a UFL expression/form as being a kernel parameter. Firedrake `Function` and `Constant` would then inherit from this.

Reply:
This conversely means that `Constant(1)*dx` needs one to define a domain on `dx`.

Reply:
> The reason we have done this is that a UFL Constant is what Firedrake calls a Function in R.

This is not quite true because you can't have a `Constant` as an unknown (for a solution variable, etc..)